state/txindex/kv/kv.go

```diff
@@ -42,6 +48,240 @@ type TxIndex struct {
+	}
+	return int64(len(affectedHeights)), retainHeight, err
+}
+
+func (txi *TxIndex) PruneOld(retainHeight int64) (int64, int64, error) {
+	// Returns numPruned, newRetainHeight, err
+	// numPruned: the number of heights pruned. E.x. if heights {1, 3, 7} were pruned, numPruned == 3
+	// newRetainHeight: new retain height after pruning
+	// err: error
+
+	lastRetainHeight, err := txi.getIndexerRetainHeight()
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to look up last block indexer retain height: %w", err)
+	}
+	if lastRetainHeight == 0 {
+		lastRetainHeight = 1
+	}
+
+	ctx := context.Background()
+	results, err := txi.Search(ctx, query.MustCompile(
+		fmt.Sprintf("tx.height < %d AND tx.height >= %d", retainHeight, lastRetainHeight)))
+	if err != nil {
+		return 0, lastRetainHeight, err
+	}
+	if len(results) == 0 {
+		return 0, lastRetainHeight, nil
+	}
+
+	batch := txi.store.NewBatch()
+	closeBatch := func(batch dbm.Batch) {
+		err := batch.Close()
+		if err != nil {
+			txi.log.Error(fmt.Sprintf("Error when closing tx indexer pruning batch: %v", err))
+		}
+	}
+	defer closeBatch(batch)
+	pruned := uint64(0)
+	flush := func(batch dbm.Batch) error {
+		err := batch.WriteSync()
+		if err != nil {
+			return fmt.Errorf("failed to flush tx indexer pruning batch %w", err)
+		}
+		err = batch.Close()
+		if err != nil {
+			txi.log.Error(fmt.Sprintf("Error when closing tx indexer pruning batch: %v", err))
+		}
+		return nil
+	}
+
+	sort.Sort(TxResultByHeight(results))
+	numHeightsBatchPruned := int64(0)                     // number of heights pruned if counting batched
+	currentBatchRetainedHeight := results[0].Height       // height retained if counting batched
+	numHeightsPersistentlyPruned := int64(0)              // number of heights pruned persistently
+	currentPersistentlyRetainedHeight := lastRetainHeight // height retained persistently
+	for i, result := range results {
+		errDeleteResult := txi.deleteResult(result, batch)
+		if errDeleteResult != nil {
+			// If we crashed in the middle of pruning the height,
+			// we assume this height is retained
+			errSetLastRetainHeight := txi.setIndexerRetainHeight(result.Height, batch)
+			if errSetLastRetainHeight != nil {
+				return 0, lastRetainHeight, fmt.Errorf("error setting last retain height '%v' while handling result deletion error '%v' for tx indexer", errSetLastRetainHeight, errDeleteResult)
+			}
+			errWriteBatch := batch.WriteSync()
+			if errWriteBatch != nil {
+				return 0, lastRetainHeight, fmt.Errorf("error writing tx indexer batch '%v' while handling result deletion error '%v'", errWriteBatch, errDeleteResult)
+			}
+			return result.Height - lastRetainHeight, result.Height, errDeleteResult
+		}
+		if i == len(results)-1 || results[i+1].Height > result.Height {
+			numHeightsBatchPruned++
+			currentBatchRetainedHeight = result.Height + 1
+		}
+		// flush every 1000 blocks to avoid batches becoming too large
+		if pruned%1000 == 0 && pruned > 0 {
```

Comment:
Variable `pruned` is never incremented

Reply:
Fixed, I think this was leftover from the copy paste he built the code on. I replaced it with `numHeightsBatchPruned` and added logic to reset `numHeightsBatchPruned` to 0 in the if.